src/lib/Renderer.ts

```diff
@@ -353,7 +343,127 @@ class Renderer {
       return { error: 'unsafe_redirect' };
     }
 
-    return { statusCode, headers, body, timeout, resolvedUrl };
+    return { statusCode, headers, body, resolvedUrl };
+  }
+
+  private async _processLogin(task: TaskParams): Promise<TaskResult> {
+    /* Setup */
+    const { url } = task;
+    const { context, page } = await this._newPageWithContext(task);
+
+    try {
+      await this._goto(page, url);
+    } catch (e) {
+      return { error: e.message, timeout: Boolean(e.timeout) };
+    }
+
+    const textInput = await page.$('input[type=text], input[type=email]');
+    if (!textInput) {
+      return { error: `field_not_found: input[type=text], input[type=email]` };
+    }
+    await textInput!.type(task.login!.username);
+
+    let passwordInput = await page.$('input[type=password]');
+    if (!passwordInput) {
+      console.log('2 step login: validating username...');
+      try {
+        await Promise.all([
+          // page.waitForNavigation(), // Doesn't work with Okta for example, it's JS based
+          page.waitForSelector('input[type=password]', {
+            timeout: TIMEOUT,
+          }),
+          textInput!.press('Enter'),
+        ]);
+      } catch (err) {
+        console.log('Found no password input on the page');
+        const body = await this._renderBody(page, new URL(page.url()));
+        return { error: err.message, body };
+      }
+    }
+
+    console.log(`2 step login: navigated to ${page.url()}`);
+    passwordInput = await page.$('input[type=password]');
+    console.log('Logging in...');
+    await passwordInput!.type(task.login!.password);
+    let loginResponse;
+    try {
+      const [navigationResponse] = await Promise.all([
+        page.waitForNavigation({ timeout: TIMEOUT }),
+        await passwordInput!.press('Enter'),
+      ]);
+      loginResponse = navigationResponse;
+    } catch (err) {
+      console.log(`Error while logging in: ${err.message} (url=${page.url()})`);
+      const body = await this._renderBody(page, new URL(page.url()));
+      return { error: err.message, body };
+    }
+
+    if (!loginResponse) {
+      console.log(`Got no login response (url=${page.url()})`);
+      const body = await this._renderBody(page, new URL(page.url()));
+      return { error: 'no_response', body };
```

Comment:
is it possible to render anything if we did not get any response?

Reply:
Yes, `page.waitForNavigation()` expects the page to reload, so if the form is validated in javascript, or if the request was done in JS but didn't succeed and we stayed on the same form with an error message, we will in theory render this page with this error. (in practice, for Okta, it was rendering the initial page, I don't know why and I can't do many tries before getting blocked. But it was rendering something anyway)

Reply:
okay 👌🏻